caffe2/operators/spatial_batch_norm_op.h

```diff
@@ -11,69 +16,377 @@ template <class Context>
+      saved_rstd->Resize(C);
+      T* saved_mean_data = saved_mean->template mutable_data<T>();
+      T* saved_rstd_data = saved_rstd->template mutable_data<T>();
+      auto* running_mean = Output(RUNNING_MEAN);
+      auto* running_var = Output(RUNNING_VAR);
+      if (running_mean->size() != C) {
+        running_mean->Resize(C);
+        math::Set<T, Context>(
+            C, T(0), running_mean->template mutable_data<T>(), &context_);
+      }
+      if (running_var->size() != C) {
+        running_var->Resize(C);
+        math::Set<T, Context>(
+            C, T(0), running_var->template mutable_data<T>(), &context_);
+      }
+      T* running_mean_data = running_mean->template mutable_data<T>();
+      T* running_var_data = running_var->template mutable_data<T>();
+      if (N == 0) {
+        math::Set<T, Context>(C, T(0), saved_mean_data, &context_);
+        math::Set<T, Context>(C, T(0), saved_rstd_data, &context_);
+        return true;
+      }
+      if (num_batches_ > 1) {
+        const auto& batch_mean_sum = Input(BATCH_MEAN_SUM);
+        const auto& batch_var_sum = Input(BATCH_VAR_SUM);
+        CAFFE_ENFORCE_EQ(batch_mean_sum.size(), C);
+        CAFFE_ENFORCE_EQ(batch_var_sum.size(), C);
+        ComputeBatchMoments<T>(
+            N,
+            C,
+            HxW,
+            batch_mean_sum.template data<T>(),
+            batch_var_sum.template data<T>(),
+            saved_mean_data,
+            saved_rstd_data);
+      } else {
+        if (order_ == StorageOrder::NCHW) {
+          const std::array<int, 3> dims = {N, C, HxW};
+          const std::array<int, 3> axes = {0, 2};
+          math::Moments<T, Context>(
+              3,
+              dims.data(),
+              2,
+              axes.data(),
+              X_data,
+              saved_mean_data,
+              saved_rstd_data,
+              &context_);
+        } else {
+          const std::array<int, 2> dims = {N * HxW, C};
+          const int axis = 0;
+          math::Moments<T, Context>(
+              2,
+              dims.data(),
+              1,
+              &axis,
+              X_data,
+              saved_mean_data,
+              saved_rstd_data,
+              &context_);
+        }
+      }
+      ComputeRunningMomentsAndFusedParam<T>(
+          C,
+          scale_data,
+          bias_data,
+          saved_mean_data,
+          saved_rstd_data,
+          running_mean_data,
+          running_var_data,
+          saved_rstd_data,
+          alpha_data,
+          beta_data);
+    }
+    if (order_ == StorageOrder::NCHW) {
+      math::AffineChannel<T, Context, StorageOrder::NCHW>(
+          N, C, HxW, X_data, alpha_data, beta_data, Y_data, &context_);
+    } else {
+      math::AffineChannel<T, Context, StorageOrder::NHWC>(
+          N, C, HxW, X_data, alpha_data, beta_data, Y_data, &context_);
+    }
+
     return true;
   }
 
  protected:
-  bool is_test_;
+  template <typename T>
+  void ComputeFusedParam(
```

Comment:
why we call it fusedParam?

Reply:
for the Forward pass, the actual formula is Y = (X - mu) / std * scale + bias. In order to reduce the number of computation, we fused it to alpha = scale / std, beta = bias - mu / std * scale. Then the formula is Y = alpha * X + beta. Here we are computing alpha and beta which is the fused param for the forward pass.